### ISSUE TITLE
Invalid Packet Length when connecting with Filezilla

### ISSUE THREAD

**Issue by blargity** [mscdex/ssh2-streams#54]
Hello,

I'm currently working on implementing an SFTP server using this project: https://github.com/BriteVerify/node-sftp-server

Whenever I connect with the command line `sftp` client everything works, but when I try to connect with Filezilla, I get this:

```
events.js:160
      throw er; // Unhandled 'error' event
      ^

Error: Bad packet length
    at SSH2Stream._transform (/Users/kevin/Development/projects/open-source/node-sftp-server/node_modules/ssh2/node_modules/ssh2-streams/lib/ssh.js:406:25)
    at SSH2Stream.Transform._read (_stream_transform.js:167:10)
    at SSH2Stream._read (/Users/kevin/Development/projects/open-source/node-sftp-server/node_modules/ssh2/node_modules/ssh2-streams/lib/ssh.js:212:15)
    at SSH2Stream.Transform._write (_stream_transform.js:155:12)
    at doWrite (_stream_writable.js:328:12)
    at writeOrBuffer (_stream_writable.js:314:5)
    at SSH2Stream.Writable.write (_stream_writable.js:241:11)
    at Socket.ondata (_stream_readable.js:555:20)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
    at Socket.Readable.push (_stream_readable.js:134:10)
    at TCP.onread (net.js:548:20)
```

It's entirely possible that Filezilla is sending some dodgy data, however it'd be good to get into the detail of what's going on and see if it's really something they're doing wrong or if it's an error in the SSH2 stream.

I'm currently using:
OS: Mac OS X 10.12.1
Node: v6.8.0
Filezilla: 3.22.1
Authentication: Public Key
Algo: ssh-rsa

There was originally [a bug filed with the downstream package](https://github.com/BriteVerify/node-sftp-server/issues/6) but it's looking like it isn't something they're going to be able to solve because it's coming from the SSH2 stream.

Is there any other info I could give that would help debug this?

**Comment by mscdex** [comment]
Can you set `debug: console.log` in your connection config object and post the resulting output?

**Comment by mscdex** [comment]
FWIW I just created a simple sftp server and connected successfully and showed a directory listing with FileZilla Client 3.21 (the latest available from my distro -- compiling from source was a huge pain and I gave up) on Linux.

**Comment by joeyhub** [comment]
I had a lot of issues like this (including this error message) getting FileZilla to work and they were pretty much all either me misinterpreting the SFTP implementation, bugs or FileZilla quirks (for example whenever it tries to read a file it floods with reads before waiting for responses even with offsets over the length of the file).

I highly suspect this is one of those categories of errors.

I recommend to implement like this:

1. Use more than one SFTP client for testing because different SFTP clients will have different quirks. You can also turn on debugging modes with these clients to see under the hood.
2. You can use wireshark and other debuggers (strace, etc). This is difficult though on an encrypted stream. Instead of packet sniffing with keys to decrypt, you can also edit the library to be verbose about writes and reads so you can check the binary data against the SFTP protocol. Just only edit in testing and delete after or delete and reinstall, etc.
3. Don't be afraid to read the source for of this library and anything else such as FileZilla.
4. Use debuggers such as node-inspector as well as adding verbosity where ever you have uncertainty to see what you actually get.
5. Connect to other normal SFTP servers (IE the one you get with openssh in linux and subsystem) using clients with debug modes. This makes it really easy. You just see what was different between their's and yours to see where the error likely is.

You've already hit the first one. I used the same two clients and found that both do things in different ways. This can be for reasons such as: The specifications are not clear or are ambiguous. Performance related reasons. Because of this they can do things in ways that are not immediately straight forward or expected. Different client options and defaults. Different tolerances for various errors.

Basically you will likely have to read specs, docs carefully, implementations and do debugging to perfect an SFTP implementation. This lib is fairly low level and doesn't work out of the box that well.

It's not only in regards to bugs. For example I had a messed up really slow implementation (IE, I didn't know name could take an array so instead I had loads of readdirs with big RTT which meant large directories took ages to list).

Simplified check list of what you'll need to make a good SFTP implementation...

1. SFTP specifications (RFCs). Print them out if you can.
2. Clients (download links).
3. Source links (either checked out, downloaded or online browse).
4. Debugging tools from npm or those included with newer versions of node.
